### PR TITLE
[SYCL-MLIR] Adopt upstream comments on transformation for nested operations

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Attempted to upstream https://github.com/intel/llvm/pull/8413 in https://reviews.llvm.org/D144648. 
The reviewer suggested to modify the pipeline instead.

This PR contains:
1. Revert https://github.com/intel/llvm/pull/8413
2. Modify pipeline to run `AffineScalarReplacementPass` both for `func.func` directly under `builtin.module` and `func.func` nested in another region, e.g., `gpu.module`.

Please do not squash and merge this PR.

### CODE REVIEW

File: polygeist/tools/cgeist/driver.cc [hunk 1]
```diff
@@ -287,8 +287,6 @@ static int canonicalize(mlir::MLIRContext &Ctx,
   OptPM.addPass(polygeist::createLoopRestructurePass());
   OptPM.addPass(polygeist::replaceAffineCFGPass());
   OptPM.addPass(mlir::createCanonicalizerPass(CanonicalizerConfig, {}, {}));
-  if (ScalarReplacement)
-    PM.addNestedPass<func::FuncOp>(mlir::createAffineScalarReplacementPass());
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/driver.cc:291]: This is not needed and causes confusion, as we were simply adding two `AffineScalarReplacementPass` one right after another one. Notice that it was adding to `PM`, so it is added after all passes added in `OptPM`.

**Review by victor-eds — APPROVED** [review]
LGTM

File: polygeist/tools/cgeist/driver.cc [hunk 2]
```diff
@@ -305,8 +303,16 @@ static int canonicalize(mlir::MLIRContext &Ctx,
...
-    if (ScalarReplacement)
+    if (ScalarReplacement) {
+      // Perform AffineScalarReplacementPass on func::FuncOp nested in another
+      // region, e.g., gpu.module.
+      mlir::OpPassManager &nestPM = PM.nestAny();
+      nestPM.addNestedPass<func::FuncOp>(
+          mlir::createAffineScalarReplacementPass());
+      // Perform AffineScalarReplacementPass on func::FuncOp directly under
+      // builtin.module.
       PM.addNestedPass<func::FuncOp>(mlir::createAffineScalarReplacementPass());
```
Comment (etiotto) [inline polygeist/tools/cgeist/driver.cc:314]: We discussed this offline, would make sense to create a lambda function to make it easier in the future to add more function passes that can operate on a `func::FuncOp`  whether it is nested directly into a builtin.module or a gpu.module.

**Review by etiotto — CHANGES_REQUESTED** [review]

File: polygeist/tools/cgeist/driver.cc (hunk 2 above)
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/driver.cc]: done.

**Review by etiotto — APPROVED** [review]
LGTM - the polygeist transformation pipeline should likely move the affine scalar replacement in the `optimize` method but that is for a future PR.

**Pull request merged**